File: src/main/java/com/better/alarm/configuration/Container.java
package com.better.alarm.configuration;

import android.app.NotificationManager;
import android.content.Context;
import android.content.SharedPreferences;
import android.media.AudioManager;
import android.os.PowerManager;
import android.os.Vibrator;
import android.telephony.TelephonyManager;

import com.better.alarm.interfaces.IAlarmsManager;
import com.better.alarm.logger.Logger;
import com.better.alarm.model.Alarms;
import com.better.alarm.wakelock.WakeLockManager;
import com.f2prateek.rx.preferences2.RxSharedPreferences;

import org.immutables.value.Value;

/**
 * Created by Yuriy on 09.08.2017.
 */
@Value.Immutable
@Value.Style(stagedBuilder = true)
public abstract class Container {
    public abstract Context context();

    @android.support.annotation.NonNull
    @Value.Derived
    public WakeLockManager wakeLocks() {
        return new WakeLockManager(logger(), powerManager());
    }

    @android.support.annotation.NonNull
    public abstract Logger logger();

    @android.support.annotation.NonNull
    public abstract SharedPreferences sharedPreferences();

    @android.support.annotation.NonNull
    public abstract RxSharedPreferences rxPrefs();

    @android.support.annotation.NonNull
    public abstract Prefs prefs();

    @android.support.annotation.NonNull
    public abstract Store store();

    @android.support.annotation.NonNull
    public abstract Alarms rawAlarms();

    @android.support.annotation.NonNull
    @Value.Derived
    public IAlarmsManager alarms() {
        return rawAlarms();
    }

    @android.support.annotation.NonNull
    @Value.Derived
    public Vibrator vibrator() {
        return (Vibrator) context().getSystemService(Context.VIBRATOR_SERVICE);
    }

    @android.support.annotation.NonNull
    @Value.Derived
    public PowerManager powerManager() {
        return (PowerManager) context().getSystemService(Context.POWER_SERVICE);
    }

    @android.support.annotation.NonNull
    @Value.Derived
    public TelephonyManager telephonyManager() {
        return (TelephonyManager) context().getSystemService(Context.TELEPHONY_SERVICE);
    }

    @android.support.annotation.NonNull
    @Value.Derived
    public NotificationManager notificationManager() {
        return (NotificationManager) context().getSystemService(Context.NOTIFICATION_SERVICE);
    }

    @android.support.annotation.NonNull
    @Value.Derived
    public AudioManager audioManager() {
        return (AudioManager) context().getSystemService(Context.AUDIO_SERVICE);
    }
}

File: app/src/main/java/com/better/alarm/configuration/Container.kt
package com.better.alarm.configuration

import android.app.AlarmManager
import android.app.NotificationManager
import android.content.Context
import android.media.AudioManager
import android.os.PowerManager
import android.os.Vibrator
import android.telephony.TelephonyManager
import android.text.format.DateFormat
import android.util.Log
import com.better.alarm.alert.BackgroundNotifications
import com.better.alarm.background.AlertServicePusher
import com.better.alarm.background.KlaxonPlugin
import com.better.alarm.background.PlayerWrapper
import com.better.alarm.bugreports.BugReporter
import com.better.alarm.interfaces.IAlarmsManager
import com.better.alarm.logger.Logger
import com.better.alarm.logger.LoggerFactory
import com.better.alarm.logger.StringUtils
import com.better.alarm.logger.loggerModule
import com.better.alarm.model.AlarmCore
import com.better.alarm.model.AlarmSetter
import com.better.alarm.model.AlarmStateNotifier
import com.better.alarm.model.Alarms
import com.better.alarm.model.AlarmsRepository
import com.better.alarm.model.AlarmsScheduler
import com.better.alarm.model.Calendars
import com.better.alarm.model.IAlarmsScheduler
import com.better.alarm.persistance.DataStoreAlarmsRepository
import com.better.alarm.persistance.DatabaseQuery
import com.better.alarm.persistance.DatastoreMigration
import com.better.alarm.persistance.SQLiteDatabaseQuery
import com.better.alarm.presenter.AlarmsListActivity
import com.better.alarm.presenter.DynamicThemeHandler
import com.better.alarm.presenter.ScheduledReceiver
import com.better.alarm.presenter.ToastPresenter
import com.better.alarm.stores.SharedRxDataStoreFactory
import com.better.alarm.util.Optional
import com.better.alarm.wakelock.WakeLockManager
import com.better.alarm.wakelock.Wakelocks
import io.reactivex.Observable
import io.reactivex.Scheduler
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.PublishSubject
import java.io.File
import java.util.*
import kotlinx.coroutines.Dispatchers
import org.koin.core.Koin
import org.koin.core.context.loadKoinModules
import org.koin.core.context.startKoin
import org.koin.core.qualifier.named
import org.koin.core.scope.Scope
import org.koin.dsl.binds
import org.koin.dsl.module


fun Scope.logger(tag: String): Logger {
  return get<LoggerFactory>().createLogger(tag)
}

fun Koin.logger(tag: String): Logger {
  return get<LoggerFactory>().createLogger(tag)
}

fun startKoin(context: Context): Koin {
  // The following line triggers the initialization of ACRA

  val module = module {
    single<DynamicThemeHandler> { DynamicThemeHandler(get()) }
    single<BugReporter> { BugReporter(logger("BugReporter"), context) }
    factory<Context> { context }
    factory(named("dateFormatOverride")) { "none" }
    factory<Single<Boolean>>(named("dateFormat")) {
      Single.fromCallable {
        get<String>(named("dateFormatOverride")).let { if (it == "none") null else it.toBoolean() }
            ?: DateFormat.is24HourFormat(context)
      }
    }

    single<Prefs> {
      val factory = SharedRxDataStoreFactory.create(get(), logger("preferences"))
      Prefs.create(get(named("dateFormat")), factory)
    }

    single<Store> {
      Store(
          alarmsSubject = BehaviorSubject.createDefault(ArrayList()),
          next = BehaviorSubject.createDefault<Optional<Store.Next>>(Optional.absent()),
          sets = PublishSubject.create(),
          events = PublishSubject.create())
    }

    factory { get<Context>().getSystemService(Context.ALARM_SERVICE) as AlarmManager }
    single<AlarmSetter> { AlarmSetter.AlarmSetterImpl(get(), get()) }

    factory { Calendars { Calendar.getInstance() } }
    single<AlarmsScheduler> {
      AlarmsScheduler(get()
//          , logger("AlarmsScheduler"), get(), get(), get()
      )
    }
    factory<IAlarmsScheduler> { get<AlarmsScheduler>() }
    single<AlarmCore.IStateNotifier> { AlarmStateNotifier(get()) }
    single<AlarmsRepository> {
      DataStoreAlarmsRepository(
          datastoreDir = get(named("datastore")),
          logger = logger("DataStoreAlarmsRepository"),
          ioDispatcher = Dispatchers.IO,
      )
    }
    single(named("datastore")) { File(get<Context>().applicationContext.filesDir, "datastore") }
    factory { get<Context>().contentResolver }
    single<DatabaseQuery> { SQLiteDatabaseQuery(get()) }
    single { Alarms(get(), get(), get(), get(), get(), get(), logger("Alarms"), get()) } binds
        arrayOf(IAlarmsManager::class, DatastoreMigration::class)
    single { ScheduledReceiver(get(), get(), get(), get()) }
    single { ToastPresenter(get(), get()) }
    single { AlertServicePusher(get(), get(), get(), logger("AlertServicePusher")) }
    single { BackgroundNotifications(get(), get(), get(), get(), get()) }
    factory<Wakelocks> { get<WakeLockManager>() }
    factory<Scheduler> { AndroidSchedulers.mainThread() }
    single<WakeLockManager> { WakeLockManager(logger("WakeLockManager"), get()) }
    factory { get<Context>().getSystemService(Context.VIBRATOR_SERVICE) as Vibrator }
    factory { get<Context>().getSystemService(Context.POWER_SERVICE) as PowerManager }
    factory { get<Context>().getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager }
    factory { get<Context>().getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager }
    factory { get<Context>().getSystemService(Context.AUDIO_SERVICE) as AudioManager }
    factory { get<Context>().resources }

    factory(named("volumePreferenceDemo")) {
      KlaxonPlugin(
          log = logger("VolumePreference"),
          playerFactory = { PlayerWrapper(get(), get(), logger("VolumePreference")) },
          prealarmVolume = get<Prefs>().preAlarmVolume.observe(),
          fadeInTimeInMillis = Observable.just(100),
          inCall = Observable.just(false),
          scheduler = get())
    }
  }

  return startKoin {
        modules(module)
        modules(AlarmsListActivity.uiStoreModule)
        modules(loggerModule())
      }
      .koin
}

fun overrideIs24hoursFormatOverride(is24hours: Boolean) {
  loadKoinModules(
      module =
          module(override = true) { factory(named("dateFormatOverride")) { is24hours.toString() } })
}
